### PR TITLE
[direnv] use new TF bucket prefix method

### LINKED ISSUE #401: Add `TF_BUCKET_PREFIX_FORMAT`
## what
Add `TF_BUCKET_PREFIX_FORMAT`

## why
* we can have one format be `pwd` and other format be `basename` or `basename-pwd` 
* `basename` is not sufficiently disambiguated when working with multi-region (e.g. `conf/us-west-2/vpc`)

## references
* https://github.com/cloudposse/geodesic/blob/master/rootfs/etc/direnv/rc.d/terraform#L16

### PR DESCRIPTION
**Pull request by raehik**
## what
* `TF_BUCKET_PREFIX_FORMAT` selects the format to use for setting the TF
remote state bucket prefix/key: the original `$(basename $(pwd))`
leaf-only format, or the new full path format.

## why

Old format uses leaf directory only, which could lead to unexpected
behaviour if you used subdirectories for structuring:

    /conf/us-east-1/vpc    -> s3://<bucket>/vpc
    /conf/eu-central-2/vpc -> s3://<bucket>/vpc

New format uses full path:

    /conf/us-east-1/vpc    -> s3://<bucket>/us-east-1/vpc
    /conf/eu-central-2/vpc -> s3://<bucket>/eu-central-2/vpc

## references
* closes #401

### CODE REVIEW

**Comment by raehik** [comment]
related issue #401

**Comment by osterman** [comment]
@raehik LGTM

**Comment by raehik** [comment]
would still like to document this somewhere, not sure where is best - don't want to place implementation details in `docs/design.md`

**Comment by raehik** [comment]
Seems to work from a quick build:

```bash
mkdir -p /conf/1/2/3
export TF_BUCKET_PREFIX_FORMAT=pwd
cd /conf/1/2/3
echo $TF_BUCKET_PREFIX # = 1/2/3
export TF_BUCKET_PREFIX_FORMAT=basename-pwd
echo $TF_BUCKET_PREFIX # = 3
export TF_BUCKET_PREFIX_FORMAT= # default
echo $TF_BUCKET_PREFIX # = 1/2/3 (new is default)
```

**Comment by raehik** [comment]
(oh, assuming `echo "use terraform" > /conf/1/2/3/.envrc`)

File: docs/design.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ We leverage as many semantics of the linux shell as we can to make the experienc
 * `/etc/motd` is the current "Message of the Day"
 * `/mnt/local` is where we house the local state (like your temporary AWS credentials)
 * `/mnt/remote` is where we mount the S3 bucket for cluster state; these files are never written to disk and only kept in memory for security
+* `/conf` is where all module configuration is stored. Directory names are mapped directly to a bucket prefix, including subdirectories.
```
Comment (osterman) [inline docs/design.md]:
```suggestion
* `/conf` is where all configurations belong. For example, stick your terraform module invocations in this directory. When using `terraform`, the directory names are mapped directly to a bucket prefix for terraform state which including subdirectories (e.g. `/conf/us-west-2/vpc` will map to a state folder prefix of 'us-west-2/vpc`).
```

**Comment by osterman** [comment]
@raehik is this ready for review?

File: docs/design.md (hunk 1 above)
Comment (raehik) [reply docs/design.md]: thanks, clearer & with an example - added

**Comment by raehik** [comment]
@osterman yep, good for review :)

File: docs/design.md [hunk 2]
```diff
@@ -29,6 +29,7 @@ We leverage as many semantics of the linux shell as we can to make the experienc
 * `/etc/motd` is the current "Message of the Day"
 * `/mnt/local` is where we house the local state (like your temporary AWS credentials)
 * `/mnt/remote` is where we mount the S3 bucket for cluster state; these files are never written to disk and only kept in memory for security
+* `/conf` is where all configurations belong. For example, stick your terraform module invocations in this directory. When using `terraform`, the directory names are mapped directly to a bucket prefix for terraform state which include subdirectories (e.g. `/conf/us-west-2/vpc` will map to a state folder prefix of 'us-west-2/vpc`).
```
Comment (osterman) [inline docs/design.md]:
```suggestion
* `/conf` is where all configurations belong. For example, stick your terraform module invocations in this directory. When using `terraform`, the directory names are mapped directly to a bucket prefix for terraform state which include subdirectories (e.g. `/conf/us-west-2/vpc` will map to a state folder prefix of `us-west-2/vpc`).
```

**Comment by osterman** [comment]
Thanks @raehik for the contribution! Keep 'em coming! =)

**Pull request merged**